mlir/lib/Dialect/Rise/EDSC/Builders.cpp

```diff
@@ -0,0 +1,665 @@
+      FunType::get(context, FunType::get(context, s, t),
+                   FunType::get(context, ArrayType::get(context, n, s),
+                                ArrayType::get(context, n, t)));
+
+  return ValueBuilder<MapSeqOp>(
+      mapType, NatAttr::get(context, n), DataTypeAttr::get(context, s),
+      DataTypeAttr::get(context, t), StringAttr::get(lowerTo, context));
+}
+
+Value mlir::edsc::op::mapSeq(StringRef lowerTo, Nat n, DataType s, DataType t,
+                             Value lambda, Value array) {
+  MLIRContext *context = ScopedContext::getContext();
+  Value mapSeq = mlir::edsc::op::mapSeq(lowerTo, n, s, t);
+  return ValueBuilder<ApplyOp>(ArrayType::get(context, n, t), mapSeq,
+                               ValueRange{lambda, array});
+}
+
+Value mlir::edsc::op::mapSeq(StringRef lowerTo, DataType s, DataType t,
+                             Value lambda, Value array) {
+  ArrayType arrayType = array.getType().dyn_cast<ArrayType>();
+  return mapSeq(lowerTo, arrayType.getSize(), s, t, lambda, array);
+}
+
+Value mlir::edsc::op::mapSeq(StringRef lowerTo, DataType s, DataType t,
+                             function_ref<Value(BlockArgument)> bodyBuilder,
+                             Value array) {
+  return mapSeq(lowerTo, s, t, lambda1(funtype(s, t), bodyBuilder), array);
+}
+
+Value mlir::edsc::op::mapSeq(StringRef lowerTo, DataType t,
+                             function_ref<Value(BlockArgument)> bodyBuilder,
+                             Value array) {
+  ArrayType arrayType = array.getType().dyn_cast<ArrayType>();
+  return mapSeq(lowerTo, arrayType.getElementType(), t, bodyBuilder, array);
+}
+
+Value mlir::edsc::op::mapSeq(DataType t,
+                             function_ref<Value(BlockArgument)> bodyBuilder,
+                             Value array) {
+  ArrayType arrayType = array.getType().dyn_cast<ArrayType>();
+  return mapSeq("loop", arrayType.getElementType(), t, bodyBuilder, array);
+}
+
+Value mlir::edsc::op::reduceSeq(StringRef lowerTo, Nat n, DataType s,
+                                DataType t) {
+  MLIRContext *context = ScopedContext::getContext();
+  FunType reduceType = FunType::get(
+      context, FunType::get(context, s, FunType::get(context, t, t)),
+      FunType::get(context, t,
+                   FunType::get(context, ArrayType::get(context, n, s), t)));
+  return ValueBuilder<ReduceSeqOp>(
+      reduceType, NatAttr::get(context, n), DataTypeAttr::get(context, s),
+      DataTypeAttr::get(context, t), StringAttr::get(lowerTo, context));
+}
+
+Value mlir::edsc::op::reduceSeq(StringRef lowerTo, Nat n, DataType s,
+                                DataType t, Value lambda, Value initializer,
+                                Value array) {
+  Value reduceOp = reduceSeq(lowerTo, n, s, t);
+  return ValueBuilder<ApplyOp>(t, reduceOp,
+                               ValueRange{lambda, initializer, array});
+}
+
+// something fishy here! Type is not correct. It is scalar -> scalar instead of
```

Comment:
Is this comment still relevant? If so, we should investigate.

Reply:
Comment is left from a bigger debugging session. Will be removed in next commit